solr/core/src/java/org/apache/solr/servlet/ServletUtils.java

```diff
@@ -216,16 +217,19 @@ static void rateLimitRequest(
       }
 
       if (!accepted) {
-        String errorMessage =
-            "Too many requests for this request type."
-                + "Please try after some time or increase the quota for this request type";
-
-        response.sendError(429, errorMessage);
+        response.sendError(ErrorCode.TOO_MANY_REQUESTS.code, RateLimitManager.ERROR_MESSAGE);
+        return;
       }
+      needsCleanup = false;
       // todo: this shouldn't be required, tracing and rate limiting should be independently
       // composable
       traceHttpRequestExecution2(request, response, limitedExecution, trace);
     } finally {
+      if (needsCleanup) {
+        consumeInputFully(request, response);
+        SolrRequestInfo.reset();
+        SolrRequestParsers.cleanupMultipartFiles(request);
```

Comment:
This looks wrong.... we won't reset SolrRequestInfo sometimes?  Why not?  Not doing so is dangerous; messing up access to SolrRequestInfo for future requests on the same thread.

And structurally this doesn't look right.  We shouldn't be adding additional callers to some of these methods.  Control flow should be well balanced logical and avoid calling methods from multiple spots under different conditions.  It's not that it's wrong necessarily but is hard to understand and reason about.

Reply:
thank you for taking a look. I agree with the control flow idea 100%. it is very difficult to work around code that is so poorly set in place.
I introduced this flag because ` traceHttpRequestExecution2` does resource cleanup "sometimes", take a look (I actually think the 'else' block in there is dead code, but it's not relevant to this PR). 
ideally this entire block "rate limiter method calling into trace method calling into mystery method" would be wrapped into a try/finally with correct cleanup but it's not so I added this flag which was "best effort cleanup" in case the trace method does not get called.
Also, is there an actual test that verifies correct resource cleanup? I am happy to extend it with this change.

If you have stronger opinions (proposing a proper refactoring) and are good for a review, I am happy to dive deeper, but so far there are not too many reviewers available.

Reply:
I'm happy to review around this code (here I am) so restructuring I think is necessary first.  @gus-asf might -- he brought about I'd current structure.

Reply:
thank you for offering. I will try to clean this up and at the same time reduce collateral changes as much as possible.